File: src/stmt.cpp
#include "stmt.h"
#include "environment.h"
#include "error.h"
#include "function.h"
#include "interpreter.h"
#include "object.h"
#include <iostream>
#include <vector>

void ExprStmt::execute(Interpreter *interpreter) {
  expr->evaluate(interpreter);
}

void PrintStmt::execute(Interpreter *interpreter) {
  Object value = expr->evaluate(interpreter);
  std::cout << value << std::endl;
}

void VarDecl::execute(Interpreter *interpreter) {
  Object value = Object(); // if the value does not have an
                           // initializer, its value will be null
  if (initializer != nullptr) {
    value = initializer->evaluate(interpreter);
  }

  interpreter->get_current()->define(name.get_lexeme(), value);
}

void execute_block(const std::vector<Stmt *> &statements, Environment *env,
                   Interpreter *interpreter) {
  Environment *previous = interpreter->get_current();
  try {
    *interpreter->set_current() = env;
    for (Stmt *statement : statements) {
      statement->execute(interpreter);
    }
  } catch (RuntimeError &e) {
    *interpreter->set_current() = previous;
    if (dynamic_cast<ReturnException*>(&e)) {
      throw;
    }
  }
  *interpreter->set_current() = previous;
}

void Block::execute(Interpreter *interpreter) {
  Environment block_environment = Environment(interpreter->get_current());
  execute_block(statements, &block_environment, interpreter);
}

void IfStmt::execute(Interpreter *interpreter) {
  Object cond_bool = condition->evaluate(interpreter);
  // using bool() cast to check if cond is true or false
  if (static_cast<bool>(cond_bool)) {
    then_branch->execute(interpreter);
  } else if (else_branch != nullptr) {
    // only when condition == false && got an 'else' statement, will we execute
    // else statement avoid dangling pointer!
    else_branch->execute(interpreter);
  }
}

void WhileStmt::execute(Interpreter *interpreter) {
  while (static_cast<bool>(condition->evaluate(interpreter))) {
    body->execute(interpreter);
  }
}

void FuncStmt::execute(Interpreter *interpreter) {
  // FIXME: a memory leak here!
  // how do we free this space?
  Function *function = new Function(this, false, interpreter->get_current());
  interpreter->get_current()->define(name.get_lexeme(), function);
}

void ReturnStmt::execute(Interpreter *interpreter) {
  Object return_value = Object();
  if (value != nullptr) {
    return_value = value->evaluate(interpreter);
  }
  throw ReturnException(return_value);
}

void delete_stmt(Stmt *stmt) {
  if (stmt == nullptr)
    return;

  if (auto var_decl = dynamic_cast<VarDecl *>(stmt)) {
    delete_expr(var_decl->initializer);
    var_decl->initializer = nullptr;
  } else if (auto expr_stmt = dynamic_cast<ExprStmt *>(stmt)) {
    delete_expr(expr_stmt->expr);
    expr_stmt->expr = nullptr;
  } else if (auto print_stmt = dynamic_cast<PrintStmt *>(stmt)) {
    delete_expr(print_stmt->expr);
    print_stmt->expr = nullptr;
  } else if (auto block_stmt = dynamic_cast<Block *>(stmt)) {
    for (Stmt *stat : block_stmt->statements) {
      delete_stmt(stat);
      stat = nullptr;
    }
  } else if (auto if_stmt = dynamic_cast<IfStmt *>(stmt)) {
    delete_expr(if_stmt->condition);
    if_stmt->condition = nullptr;
    delete_stmt(if_stmt->then_branch);
    if_stmt->then_branch = nullptr;
    delete_stmt(if_stmt->else_branch);
    if_stmt->else_branch = nullptr;
  } else if (auto while_stmt = dynamic_cast<WhileStmt *>(stmt)) {
    delete_expr(while_stmt->condition);
    while_stmt->condition = nullptr;
    delete_stmt(while_stmt->body);
    while_stmt->body = nullptr;
  }
  delete stmt;
  stmt = nullptr;
}

File: src/stmt.rs
use crate::expr::*;
use crate::token::Token;

#[derive(Debug, PartialEq, Clone)]
pub enum Stmt {
    VarDecl(VarDecl),
    Expr(ExprStmt),
    Print(PrintStmt),
    Block(BlockStmt),
    If(IfStmt),
    While(WhileStmt),
    Function(FunctionStmt),
    Return(ReturnStmt),
    Class(ClassStmt),
    Invalid,
}

#[derive(Debug, PartialEq, Clone)]
pub struct VarDecl {
    pub name: Token,
    // if not given a value when declaration

    // all variable will have `nil` as their default value
    pub initializer: Box<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ExprStmt {
    pub expr: Box<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct PrintStmt {
    pub expr: Box<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct BlockStmt {
    pub statements: Vec<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct IfStmt {
    pub condition: Box<Expr>,
    pub then_branch: Box<Stmt>,
    pub else_branch: Option<Box<Stmt>>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct WhileStmt {
    pub condition: Box<Expr>,
    pub body: Box<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct FunctionStmt {
    pub name: Token,
    pub params: Vec<Token>,
    pub body: Vec<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ReturnStmt {
    pub keyword: Token,
    pub value: Box<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ClassStmt {
    pub name: Token,
    pub super_class: Option<Box<Expr>>,
    pub methods: Vec<Stmt>,
}

pub trait StmtVisitor<T> {
    fn visit_var_decl(&mut self, stmt: &VarDecl) -> T;
    fn visit_expr_stmt(&mut self, stmt: &ExprStmt) -> T;
    fn visit_print_stmt(&mut self, stmt: &PrintStmt) -> T;
    fn visit_block_stmt(&mut self, stmt: &BlockStmt) -> T;
    fn visit_if_stmt(&mut self, stmt: &IfStmt) -> T;
    fn visit_while_stmt(&mut self, stmt: &WhileStmt) -> T;
    fn visit_function_stmt(&mut self, stmt: &FunctionStmt) -> T;
    fn visit_return_stmt(&mut self, stmt: &ReturnStmt) -> T;
    fn visit_class_stmt(&mut self, stmt: &ClassStmt) -> T;
}

impl Stmt {
    pub fn accept<T>(&self, visitor: &mut dyn StmtVisitor<T>) -> T {
        match self {
            Self::VarDecl(stmt) => visitor.visit_var_decl(stmt),
            Self::Expr(stmt) => visitor.visit_expr_stmt(stmt),
            Self::Print(stmt) => visitor.visit_print_stmt(stmt),
            Self::Block(stmt) => visitor.visit_block_stmt(stmt),
            Self::If(stmt) => visitor.visit_if_stmt(stmt),
            Self::While(stmt) => visitor.visit_while_stmt(stmt),
            Self::Function(stmt) => visitor.visit_function_stmt(stmt),
            Self::Return(stmt) => visitor.visit_return_stmt(stmt),
            Self::Class(stmt) => visitor.visit_class_stmt(stmt),
            Self::Invalid => panic!("Attempting to run an Invalid Statement!"),
        }
    }
}
